type-check/sum.test.ts

```diff
@@ -5,19 +5,13 @@ const { checks, check } = Test;
 
 // const res0 = sum([1, 'a']); // error
 const res1 = sum([1, 2]);
-const res2 = sum(["a", "b"]);
-const res3 = sum(toAsync([1, 2]));
-const res4 = pipe([1, 2], sum);
-const res5 = pipe(["a", "b"], sum);
-const res6 = pipe([1, 2], toAsync, sum);
-const res7 = pipe(["a", "b"], toAsync, sum);
+const res2 = sum(toAsync([1, 2]));
+const res3 = pipe([1, 2], sum);
+const res4 = pipe([1, 2], toAsync, sum);
 
 checks([
   check<typeof res1, number, Test.Pass>(),
-  check<typeof res2, string, Test.Pass>(),
+  check<typeof res2, Promise<number>, Test.Pass>(),
   check<typeof res3, Promise<number>, Test.Pass>(),
```

Comment:
`res3` should be `number` type not `Promise<number>`

Reply:
oops... if the type is `number`, does not pass type check...

<img width="579" alt="image" src="https://user-images.githubusercontent.com/10627668/188815863-36b69719-abc0-41e7-9ac2-a265aee98d37.png">


Is it needs changed this code?
```ts
function sum<A extends readonly []>(iterable: A): 0;
function sum<A extends number[]>(arr: A): number;
function sum<A extends Iterable<number>>(iterable: A): number;
function sum<A extends AsyncIterable<number>>(iterable: A): Promise<number>;

function sum<A extends Iterable<number> | AsyncIterable<number>>(
  iterable: A,
): number | Promise<number> {
  if (Array.isArray(iterable)) {
    return iterable.reduce(add, 0);
  } else if (isIterable(iterable)) {
    return reduce(add, iterable) as ReturnValueType<A>;
  } else if (isAsyncIterable(iterable)) {
    return reduce<number, number>(add, Promise.resolve(0), iterable);
  }

  throw new TypeError("'iterable' must be type of Iterable or AsyncIterable");
}
```

Reply:
I'll check it out too.

Reply:
The syntax does not seem to work properly with type inference.
```ts
function sum<A extends Iterable<number>>(iterable: A): number;
function sum<A extends AsyncIterable<number>>(iterable: A): Promise<number>;
```

Reply:
why...

Reply:
@hg-pyun 
As it is a type affected by pipe, as in the comment [here](https://github.com/marpple/FxTS/pull/169#discussion_r945309277)

I think it should be solved in the following way
```ts
function sum<A extends Iterable<number> | AsyncIterable<number>>(
  iterable: A,
): A extends Iterable<number>
  ? number
  : A extends AsyncIterable<number>
  ? Promise<number>
  : never;
```

If you do as above, you can check that type inference works.
```ts
const res0 = sum([]);
const res1 = sum([1, 2]);
const res2 = sum(toAsync([1, 2]));
const res3 = pipe([], sum);
const res4 = pipe([1, 2], sum);
const res5 = pipe([1, 2], toAsync, sum);

checks([
  check<typeof res0, number, Test.Pass>(),
  check<typeof res1, number, Test.Pass>(),
  check<typeof res2, Promise<number>, Test.Pass>(),
  check<typeof res3, number, Test.Pass>(),
  check<typeof res4, number, Test.Pass>(),
  check<typeof res5, Promise<number>, Test.Pass>(),
]);


```

Reply:
@hg-pyun  yes,  Since we need to first deliver the type signature provided to the user, the internal implementation of reduce type can proceed later.